runtime/Cpp/runtime/src/atn/SemanticContext.cpp

```diff
@@ -13,6 +17,41 @@ using namespace antlr4;
 using namespace antlr4::atn;
 using namespace antlrcpp;
 
+namespace {
+
+  struct SemanticContextHasher final {
+    size_t operator()(const SemanticContext *semanticContext) const {
+      return semanticContext->hashCode();
+    }
+  };
+
+  struct SemanticContextComparer final {
+    bool operator()(const SemanticContext *lhs, const SemanticContext *rhs) const {
+      return *lhs == *rhs;
+    }
+  };
+
+  template <typename Comparer>
+  void insertSemanticContext(const Ref<const SemanticContext> &semanticContext,
+                             std::unordered_set<const SemanticContext*, SemanticContextHasher, SemanticContextComparer> &operandSet,
+                             std::vector<Ref<const SemanticContext>> &operandList,
+                             Ref<const SemanticContext::PrecedencePredicate> &precedencePredicate) {
+    if (semanticContext != nullptr) {
+      if (is<const SemanticContext::PrecedencePredicate>(semanticContext)) {
+        if (precedencePredicate == nullptr || (Comparer{})(downCast<const SemanticContext::PrecedencePredicate*>(semanticContext.get())->precedence, precedencePredicate->precedence)) {
```

Comment:
Two typecasts? Can you down cast to the Comparer directly? Also it seems odd that you use a C-style cast here, which we are usually try to avoid.

Reply:
This was template stuff. I switched it so the comparer is passed via a parameter and no more C style looking cast, its just a operator() call now.

Reply:
its comparing the precedence value, int.